A pacing algorithm (#3630)
quicwg/base-drafts: draft-ietf-quic-recovery.md

containing only ACK frames SHOULD therefore not be paced, to avoid delaying
their delivery to the peer.

Endpoints can implement pacing as they choose. A perfectly paced sender spreads

> bytes exactly evenly over time. For a window-based congestion controller, such

Reviewer: ```suggestion
packets exactly evenly over time. For a window-based congestion controller, such
```

Given overheads, spreading bytes evenly is impractical.

as the one in this document, that rate can be computed by averaging the
congestion window over the round-trip time. Expressed as a rate in bytes:

~~~
rate = N * congestion_window / smoothed_rtt
~~~

Or, expressed as an inter-packet interval:

~~~
interval = smoothed_rtt * packet_size / congestion_window / N
~~~

Using a value for `N` that is small, but at least 1 (for example, 1.25) ensures

> that short-term variations in round-trip time or scheduler delays don't result

Reviewer: ```suggestion
that variations in round-trip time don't result
```

I'd address the scheduler delay issue separately(also, 1.25x may not be enough to fix scheduler delay) and the variations in round trip time may not be short term(ie: a buffer draining).

> in under-utilization of the congestion window.

Reviewer: ```suggestion
in under-utilization of the congestion window.  Values of 'N' larger than 1 result
in the sender being ack-clocked when the window is fully utilized and
acknowledgements arrive smoothly.
```

Reviewer: SG, I edited to no longer rely on the term ack-clocked.

max_datagram_size is the current maximum size of a datagram for the connection,
not including UDP or IP overhead.

> Endpoints can implement pacing as they choose. A perfectly paced sender spreads

Reviewer 2: I would propose moving this paragraph up one, and adding this to the start of the previous paragraph: "Practical considerations, such as packetization, scheduling delays, and computational efficiency, can cause a sender to deviate from this rate over time periods that are much shorter than a round-trip time. Sending multiple packets into the network without any delay ..."

Author: Will do.  I also noticed that that paragraph refined the initial congestion window, so I replaced the (re)definition with a section reference.

bytes exactly evenly over time. For a window-based congestion controller, such
as the one in this document, that rate can be computed by averaging the
congestion window over the round-trip time. Expressed as a rate in bytes:

~~~
rate = N * congestion_window / smoothed_rtt
~~~

Or, expressed as an inter-packet interval:

~~~
interval = smoothed_rtt * packet_size / congestion_window / N
~~~

Using a value for `N` that is small, but at least 1 (for example, 1.25) ensures
that short-term variations in round-trip time or scheduler delays don't result

> in not fully utilizing the congestion window.

Reviewer 2: ```suggestion
in under-utilization of the congestion window.
```

max_datagram_size is the current maximum size of a datagram for the connection,
not including UDP or IP overhead.

Endpoints can implement pacing as they choose, as long as the rate of bytes
sent over any interval, `t[n] - t[m]`, does not exceed a small multiple, `N`,
of maximum value of the congestion window, `congestion_window`, averaged over
the round trip time estimate, `smoothed_rtt`. Allowing for bursts produces the
following relation for the total number of bytes sent over any interval:

~~~

> sent@t[n] - sent@t[m]

Author: ```suggestion
number of bytes sent between t[m] and t[n]
```

    <= N * congestion_window * (t[m] - t[n]) / smoothed_rtt
       + min(10 * max_datagram_size,
             max(2 * max_datagram_size, 14720))
~~~

> The inclusion of the factor, `N`, that increases the rate of sending is to
> anticipate increases in sending rate as the congestion window expands. Sending
> at a higher rate also avoids having scheduling delays reduce the send rate
> below that allowed by the congestion controller. A value of 2 ensures that
> pacing is unlikely to limit what is sent any more than the congestion
> controller.

Author: ```suggestion
The inclusion of the factor, `N`, with a value of at least 1 increases the rate
of sending. The value of `N` might be increased to avoid having scheduling
delays reduce the send rate below that allowed by the congestion controller.
```

> 

Author: ```suggestion

An endpoint that does not burst packets and has even spacing between packets
might send packets spaced at intervals of:

~~~
inter-packet time = smoothed_rtt * packet_size / congestion_window / N
~~~
```

max_datagram_size is the current maximum size of a datagram for the connection,
not including UDP or IP overhead.

> Endpoints can implement pacing as they choose, as long as the rate of bytes

Reviewer 2: I think this is all too much prescription. I suggest you move this up above the previous paragraph, and lead into the burst paragraphs like so: 

"Endpoints can implement pacing as they choose. A perfectly paced sender spreads bytes exactly evenly over time. For a window-based congestion controller, such as the one in this document, that rate can be computed by averaging the congestion window over the round-trip time. More precisely, `pacing rate = N * congestion_window / smoothed_rtt`, where using a small N (for example, 1.25) protects the sender from short-term variations in round-trip time or from scheduler delays.

Further practical considerations, such as packetization, scheduling delays, and computational efficiency, can cause a sender to deviate from this rate over time periods that are much shorter than a round-trip time. Sending multiple packets into the network without any delay ..."
